### DISCUSSION TITLE
Query the same foreign table multiple times via One-to-many

### DISCUSSION THREAD

**Discussion by hgiesel** [supabase/supabase#13035] [Questions] [answered]
In the docs there's this example, to query the same table multiple times via Many-to-one:

```ts
const { data, error } = await supabase
  .from('messages')
  .select(`
    content,
    from:sender_id(name),
    to:receiver_id(name)
  `)
```

Data source:

```sql
CREATE TABLE users (
  id int8 primary key,
  name text
);

CREATE TABLE messages (
  sender_id int8 not null references users,
  receiver_id int8 not null references users,
  content text
);
```

But what if I wanted to start from users instead, and get all messages they sent/received by id, what would the query look like then? Something like this?

```ts
const { data, error } = await supabase
  .from('users')
  .select(`
    name,
    sentMessages:sender_id(id), // ???
    receivedMessages:receiver_id(id)  // ???
  `)
```

**Comment by steve-chavez** [comment, marked as answer]
You can do:

```js
const { data, error } = await supabase
  .from('users')
  .select(`
    name,
    sentMessages:messages!sender_id(*), 
    receivedMessages:messages!receiver_id(*)  
  `)
```

**Reply by devhandler** [reply]
@steve-chavez could you share which docs explain the usage of "!" syntax and examples? i searched the below and couldn't find anything
https://supabase.com/docs/reference/javascript/select

**Reply by GaryAustin1** [reply]
@devhandler 
![image](https://github.com/supabase/supabase/assets/54564956/00b8d014-910c-45d7-a8f6-36e2d90d542f)
from that page you linked.
And postgrest.org is the best resource for the database REST API:
https://postgrest.org/en/stable/references/api/resource_embedding.html?highlight=inner#top-level-filtering

**Reply by devhandler** [reply]
thank you!

**Comment by devhandler** [comment]
@hgiesel did you encounter any error when your table references foreign tables multiple times via different cols?
when i use col(foreign_col) like the example, i got an error as below:

```
code: "PGRST120"
details: "Only is null or not is null filters are allowed on embedded resources"
hint: null
message: "Bad operator on the 'profile_id' embedded resource"
```

https://github.com/orgs/supabase/discussions/16210
